### DISCUSSION TITLE
Add versioning to difficulty calculators

### DISCUSSION THREAD

**Discussion by Syriiin** [ppy/osu#13356] [Ideas]
# Proposal

I would like to propose simply adding a public `Version` property to difficulty calculators to denote the version of the difficulty calculator modules and updating it when changes are made in the difficulty calculators.
I feel this would be very helpful in keeping track of difficulty calculation updates, and especially for 3rd party applications and calculators where it is usually very difficult to determine which implementation a calculator is using.

I propose a simple semver versioning system consisting of the following parts:
MAJOR - Incremented when there is a major rework in the difficulty/pp calculation system (eg. fundamental rework).
MINOR - Incremented when there are changes that cause a change in calculated difficulty or performance values (eg. skill changes).
PATCH - Incremented when there are changes that merely refactor the calculator without affecting calculated difficulty or performance values (eg. refactors).

Following this approach, I propose we start the versioning of all calculators at `1.0.0`.

Additionally, I would propose we store this in the backend database, and propagate the difficulty and performance value versions through the osu-web API, however this is a natural progression and not a cause for immediate consideration.

## Questions

_What constitutes a "major" change?_

My opinion is that a "major" change is, for example, that of ppv1 to ppv2.
Since the fundamental method of difficulty calculation (weighted sum of strains) hasn't changed since it's initial implementation, I think this is a reasonable bar to set as a "major" change.

_How should changes to the abstract `DifficultyCalculator` be handled?_

My opinion is that changes to the abstract calculator should be merely seen as common changes to each calculator and thus should increment each of the individual calculator's relevant version numbers accordingly.

_Why start at `1.0.0` for all calculators when we consider standard to use "ppv2"?_

If we accept my thoughts on what a "major" version change is, then it might be reasonable to say we should start at `2.x.x` for the standard difficulty calculator, however the appropriate minor and patch versions would be extremely difficulty to determine; not to mention the other rulesets, where I wouldn't even know where to begin thinking.
For this reason I think it would be a good idea to simply start all the calculators at `1.0.0`.

_Why not just use the published NuGet package version or commit hash?_

While this is possible, it has very limited usefulness.
99% of the time, the NuGet package version will be bumped without any changes to the difficulty calculators.
Also, since the ruleset NuGet package versions are based on the date, they give no indication as to whether an update will result in changes to the difficulty or performance values.
Commit hashes are even less useful for obvious reasons.

**Comment by smoogipoo** [comment]
Couldn't third party applications just store the commit hash of the version they're implementing? Be it based on an osu.Game release/tag or otherwise.

**Reply by Syriiin** [reply]
I was actually just editing to write the reply to this 😅 .
Yes, but it's hardly as useful.
For example, if we wanted to use the version to determine whether we need to recalculate values, the commit hash doesn't help at all, since it is directly related to the difficulty calculators. The NuGet package version is the same story.
Having semver versioning would be the idea system for determining the validity of database values (simply recalculate if MINOR or MAJOR version update).

**Reply by Syriiin** [reply]
Additionally, alternate difficulty calculator proposals can't use the NuGet package version (since they aren't published), and the commit hash has no linearity.
It would be much more friendly to have a system that could be implemented by _any_ implementation of a difficulty calculator.

**Comment by peppy** [comment]
I think if we are attaching versions they would just follow our standard date-based versioning. Don't see much benefit in semver for a system that is 99% consumed by our infrastructure (which I manage, so is not really requiring discussion here).

**Reply by Syriiin** [reply]
The date-based versioning works too, so long as it is independent of non-difficulty calculator changes.
However, I would argue that being able to programatically identify whether a version contains changes that require recalculating existing values is a very important benefit of semver, especially for 3rd party applications.
Semver ultimately allows an additionally level of automation.

**Reply by peppy** [reply]
Yeah I dunno. We wouldn't be doing that level of automation at our end and it's such an edge case I'm not sure it's even worth consideration at this point. Just read the changelog and decide whether a reprocess is required for the time being.

**Comment by Syriiin** [comment]
For context, I am proposing this as I am working on a system to automate recalculation of live difficulty and performance values for osu!chan, which is reliant on a versioning scheme that allows programmatic detection of recalculations.

**Comment by Syriiin** [comment]
There is also a discussion to be had as to whether performance calculators should have independent versions or not, since there can be changes to performance calculators that do not require recalculation of difficulty values, I believe this would be beneficial.
